src/lib/pubkey/classic_mceliece/cmce_field_ordering.cpp

```diff
@@ -0,0 +1,322 @@
+   static_assert(sizeof(T1) <= sizeof(uint64_t) && sizeof(T2) <= sizeof(uint64_t),
+                 "Types T1 and T2 must be at most 64 bits wide");
+   if((i & k) == 0) {  // i and k do not depend on secret data
+      auto swap_required_mask = CT::Mask<uint64_t>::is_lt(a[l].first, a[i].first);
+      cond_swap_pair(swap_required_mask, a[i], a[l]);
+   } else {
+      auto swap_required_mask = CT::Mask<uint64_t>::is_gt(a[l].first, a[i].first);
+      cond_swap_pair(swap_required_mask, a[i], a[l]);
+   }
+}
+
+// Sorts a vector of pairs after the first element
+template <std::unsigned_integral T1, std::unsigned_integral T2>
+void bitonic_sort_pair(std::span<std::pair<T1, T2>> a) {
+   const size_t n = a.size();
+   BOTAN_ARG_CHECK(is_power_of_2(n), "Input vector size must be a power of 2");
+
+   for(size_t k = 2; k <= n; k *= 2) {
+      for(size_t j = k / 2; j > 0; j /= 2) {
+         for(size_t i = 0; i < n; ++i) {
+            const size_t l = i ^ j;
+            if(l > i) {
+               compare_and_swap_pair(a, i, k, l);
+            }
+         }
+      }
+   }
+}
+
+template <std::unsigned_integral T>
+T min(const T& a, const T& b) {
+   auto mask = CT::Mask<T>::is_lt(a, b);
+   return mask.select(a, b);
+}
+
+}  // namespace CMCE_CT
+
+namespace {
+template <std::unsigned_integral T1, std::unsigned_integral T2>
+std::vector<std::pair<T1, T2>> zip(std::span<const T1> vec_1, std::span<const T2> vec_2) {
+   BOTAN_ARG_CHECK(vec_1.size() == vec_2.size(), "Vectors' dimensions do not match");
+   std::vector<std::pair<T1, T2>> vec_zipped;
+   vec_zipped.reserve(vec_1.size());
+   for(size_t i = 0; i < vec_1.size(); ++i) {
+      vec_zipped.push_back(std::make_pair(vec_1[i], vec_2[i]));
+   }
+   return vec_zipped;
+}
+
+template <std::unsigned_integral T1, std::unsigned_integral T2>
+std::pair<secure_vector<T1>, secure_vector<T2>> unzip(const std::span<std::pair<T1, T2>>& vec_zipped) {
+   std::pair<secure_vector<T1>, secure_vector<T2>> res;
+
+   res.first.reserve(vec_zipped.size());
+   res.second.reserve(vec_zipped.size());
+
+   for(const auto& [elem1, elem2] : vec_zipped) {
+      res.first.push_back(elem1);
+      res.second.push_back(elem2);
+   }
+   return res;
+}
+
+/// @returns (vec[0],0), ..., (vec[n-1],n-1)
+std::vector<std::pair<uint32_t, uint16_t>> enumerate(std::span<const uint32_t> vec) {
+   std::vector<std::pair<uint32_t, uint16_t>> enumerated;
+
+   std::transform(vec.begin(), vec.end(), std::back_inserter(enumerated), [ctr = uint16_t(0)](uint32_t elem) mutable {
+      return std::make_pair(elem, ctr++);
+   });
+
+   return enumerated;
+}
+
+/**
+* @brief Create permutation pi as in (Section 8.2, Step 3).
+*/
+CmcePermutation create_pi(std::span<const uint32_t> a) {
+   auto a_pi_zipped = enumerate(a);
+   CMCE_CT::bitonic_sort_pair(std::span(a_pi_zipped));
+
+   CmcePermutation pi_sorted;
+   std::tie(a, pi_sorted.get()) = unzip(std::span(a_pi_zipped));
+
+   return pi_sorted;
+}
```

Comment:
Reminder to me: This method is messed up. We want to return the sorted `a` values too, otherwise the std::adjacent_find will not find repetitions. Also add a regression test for this.

Reply:
Good catch! I just saw we used to have `secure_vector<uint32_t>& a` as input